### PR TITLE
Botoes edit delete

### PR DESCRIPTION
**Pull request by JessicaFranke**
Botões de _delete_ e _edit_ visíveis. Por enquanto só o botão de _delete_ foi implementado, para autores e cursos.

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -8,7 +8,7 @@ const routes = require('./src/routes/routes')
 
 const app = express()
 
-const MONGO_URL = process.env.DATABASELOGIN
```
Comment (KarenF) [inline app.js]: Fica esse ae

File: mongoScript.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        title: "Operações de CRUD (Create, Read, Update, Delete)",
+        author: "Waldemar Neto",
+        type: "video",
+        url: "arVY-N0RNPQ",
+        preview: "No quinto video da serie de Node.js API testável vamos fazer as operações de CRUD da nossa API e também fazer testes para elas para garantir a funcionalidade.",
+        views: 16
+      }
+    ]
+  }]
+  })
```
Comment (KarenF) [inline mongoScript.js]: 11º mandamento: Não comitaras scripts em vão

File: views/partials/authorsList.handlebars [hunk 3]
```diff
@@ -10,12 +10,31 @@
...
+  if (confirm('Esta operação deletará todos os cursos deste autor. Você tem certeza disto?')){
+    var data = {};
+    data.id = $('#id-{{_id}}').text()
+    data.collection = 'author'
+    $.ajax('http://localhost:3000/deletePost', {
+      type: 'POST',
+      data: JSON.stringify(data),
+      contentType: 'application/json',
+      success: function() { console.log('success');},
+      error  : function() { console.log('error');}
```
Comment (KarenF) [inline views/partials/authorsList.handlebars]: precisa do ponto e virgula?

File: views/partials/authorsList.handlebars [hunk 4]
```diff
@@ -10,12 +10,31 @@
...
 </div>
+<script language="JavaScript" type="text/javascript">
+{{#allAuthors}}
+$('#deleteAuthor-{{_id}}').click(function(e) {
+  e.preventDefault();
+  if (confirm('Esta operação deletará todos os cursos deste autor. Você tem certeza disto?')){
+    var data = {};
+    data.id = $('#id-{{_id}}').text()
+    data.collection = 'author'
+    $.ajax('http://localhost:3000/deletePost', {
```
Comment (KarenF) [inline views/partials/authorsList.handlebars]: Não deveria ser URL no lugar de localhost?

File: views/partials/coursesList.handlebars [hunk 5]
```diff
@@ -11,13 +11,34 @@
...
+<script language="JavaScript" type="text/javascript">
+{{#allAuthors}}
+{{#courses}}
+$('#deleteCourse-{{_id}}').click(function(e) {
+  e.preventDefault();
+  if (confirm('Esta operação deletará todos os conteúdos deste curso. Você tem certeza disto?')){
+    var data = {};
+    data.id = $('#id-{{_id}}').text()
+    data.collection = 'courses'
+    $.ajax('http://localhost:3000/deletePost', {
```
Comment (KarenF) [inline views/partials/coursesList.handlebars]: Mesma situação da URL anterior

File: views/partials/coursesList.handlebars [hunk 6]
```diff
@@ -11,13 +11,34 @@
...
+      type: 'POST',
+      data: JSON.stringify(data),
+      contentType: 'application/json',
+      success: function() {console.log('success');},
+      error  : function() { console.log('error');}
```
Comment (KarenF) [inline views/partials/coursesList.handlebars]: ponto e virgula

File: views/partials/authorsList.handlebars (hunk 4 above)
Comment (othman853) [reply views/partials/authorsList.handlebars]: `/deletePost` já serve, `localhost:3000` vai falhar no Heroku.

Comment (othman853) [reply views/partials/authorsList.handlebars]: https://github.com/axios/axios

File: views/partials/coursesList.handlebars (hunk 5 above)
Comment (othman853) [reply views/partials/coursesList.handlebars]: https://github.com/axios/axios

File: views/layouts/main.handlebars [hunk 7]
```diff
@@ -4,6 +4,7 @@
   <meta charset="utf-8">
   <meta name="viewport" content="width=device-width, initial-scale=1">
   <link rel="stylesheet" href="/css/bulma.css">
+  <script src="//ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
```
Comment (othman853) [inline views/layouts/main.handlebars]: https://github.com/axios/axios

File: views/partials/authorsList.handlebars [hunk 8]
```diff
@@ -10,12 +10,31 @@
...
     {{/allAuthors}}
   </table>
 </div>
+<script language="JavaScript" type="text/javascript">
+{{#allAuthors}}
+$('#deleteAuthor-{{_id}}').click(function(e) {
+  e.preventDefault();
+  if (confirm('Esta operação deletará todos os cursos deste autor. Você tem certeza disto?')){
+    var data = {};
+    data.id = $('#id-{{_id}}').text()
```
Comment (othman853) [inline views/partials/authorsList.handlebars]: Isso funciona? O Handlebars consegue interpolar JavaScript?

File: views/partials/authorsList.handlebars [hunk 9]
```diff
@@ -10,12 +10,31 @@
     <tr>
       <td>{{name}}</td>
       <td>{{email}}</td>
-      <td>{{_id}}</td>
+      <td id="id-{{_id}}">{{_id}}</td>
```
Comment (othman853) [inline views/partials/authorsList.handlebars]: Isso deveria ser uma tabela mesmo?

File: views/partials/authorsList.handlebars (hunk 8 above)
Comment (othman853) [reply views/partials/authorsList.handlebars]: Isso poderia ir para um arquivo separado. Alternativamente, seria possível utilizar `forEach` para definir esse evento para todos os autores. Parece ser mais simples usar elementos relativos nesse caso do que gerar um id para cada elemento:

```html
<div class="author-data">
  <p class="author-id">asdfasdf12312313fasjh398r</p>
  <button> Remover autor </button>
</div>
```

```javascript
// Busca todos os autores do documento
var allAuthorData = document.querySelectorAll('.author-data');

// para cada um dos autores...
allAuthorData.forEach(function(authorData) {
  
  // Obtém o id de tal autor
  var authorId = authorData.querySelector('.author-id').innerText;

  //Atribui um event listener para cada botao de excluir:
  authorData.querySelector('button').addEventListener('click', function() {
    axios
      .post('/delete', {id: authorId})
      .then(function() { console.log('OK') })
      .catch(function(err) {console.log('Error', err.message})
  })
})
```

File: views/partials/authorsList.handlebars [hunk 10]
```diff
@@ -10,12 +10,31 @@
...
+        <button id="deleteAuthor-{{_id}}" type="button" name="delete">Delete</button>
         <button type="button" name="edit">Edit</button>
       </td>
     </tr>
     {{/allAuthors}}
   </table>
 </div>
+<script language="JavaScript" type="text/javascript">
+{{#allAuthors}}
+$('#deleteAuthor-{{_id}}').click(function(e) {
```
Comment (othman853) [inline views/partials/authorsList.handlebars]: Vocês não precisam de JQuery para isso.

File: app.js [hunk 11]
```diff
@@ -36,6 +36,9 @@ app.get('/course/:id', routes.course)
 
 app.get('/author/:id', routes.author)
 
+app.post('/deletePost', routes.deletePost)
```
Comment (othman853) [inline app.js]: O que é um `Post`? Não existe esse conceito no modelo de vocês até então. O que está sendo deletado nesse endpoint?

Comment (othman853) [inline app.js]: Idealmente, poderíamos utilizar `app.delete`, que é outro método http específico para representar ações de exclusão, não se precisa de nada além do id do recurso que se quer deletar, portanto, requests delete não têm corpo:

```javascript
app.delete('/author/deleteCourses/:authorId', (req, res) => {
 /* Do the magic*/
})
```

- Sobre o verbo [HTTP Delete](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/DELETE)
- Sobre [.delete do express](https://expressjs.com/en/starter/basic-routing.html)

File: views/partials/authorsList.handlebars (hunk 3 above)
Comment (othman853) [reply views/partials/authorsList.handlebars]: No caso do ponto e vírgula, é provável que continue sendo opcional. É importante lembrar que

### Esse código roda no browser

O que significa que __o JavaScript que roda aqui não é o mesmo que roda no backend pelo node!__ Isso quer dizer que alguns recursos são incompatíveis em alguns browsers, por isso, evitem funcionalidades muito avançadas se quiserem manter compatibilidade com browsers mais antigos. 

Arrow functions não são suportadas pelo `IE11`, por exemplo.

Mais informações sobre compatibilidade [nesse site bacana](https://caniuse.com/#search=arrow).

File: views/partials/coursesList.handlebars (hunk 5 above)
Comment (othman853) [reply views/partials/coursesList.handlebars]: Lembrando que 

# Não precisa de Jquery

File: views/layouts/main.handlebars (hunk 7 above)
Comment (othman853) [reply views/layouts/main.handlebars]: Salientando que

# JQuery não é necessário.

**Comment by othman853** [comment]
http://youmightnotneedjquery.com/
http://youmightnotneedjquery.com/
http://youmightnotneedjquery.com/

**Comment by othman853** [comment]
http://youmightnotneedjquery.com/
http://youmightnotneedjquery.com/
http://youmightnotneedjquery.com/
http://youmightnotneedjquery.com/
http://youmightnotneedjquery.com/
http://youmightnotneedjquery.com/
http://youmightnotneedjquery.com/
http://youmightnotneedjquery.com/
http://youmightnotneedjquery.com/
http://youmightnotneedjquery.com/

File: .gitignore [hunk 12]
```diff
@@ -63,6 +63,9 @@ test/e2e/reports/
 # css files
 public/css/
 
+# script for mongodb
+mongoScript.js
```
Comment (othman853) [inline .gitignore]: Porque esse script está no `gitignore`? Ele deveria ser um script de seed e deveria ser versionado junto com o projeto. E se o `schema` do banco mudar? O Script precisa mudar também, e ele precisa estar atualizado.

File: app.js (hunk 11 above)
Comment (othman853) [reply app.js]: Aonde está o arquivo desta rota? Ele foi commitado?

File: app.js [hunk 13]
```diff
@@ -2,6 +2,7 @@ const express = require('express')
 const path = require('path')
 const mongoose = require('mongoose')
 const bodyParser = require('body-parser')
+const axios = require('axios')
```
Comment (othman853) [inline app.js]: Não precisa do `require` no backend, vocês estão usando axios do lado do browser. 😄

File: app.js [hunk 14]
```diff
@@ -43,14 +44,21 @@ app.get('/', routes.index)
...
+
 app.get('/content/:id', routes.content)
 
 app.get('/course/:id', routes.course)
 
 app.get('/author/:id', routes.author)
 
-app.use((req, res) =>
-  res.status(404).render('404'))
+app.post('/deleteItem/:id', routes.deleteItem)
```
Comment (othman853) [inline app.js]: Essa rota segue sendo `post`. Seria possível usar `delete` aqui?

File: package.json [hunk 15]
```diff
@@ -23,6 +23,7 @@
   },
   "homepage": "https://github.com/aceleradora-TW/conhecimento-livre#readme",
   "dependencies": {
+    "axios": "^0.17.1",
```
Comment (othman853) [inline package.json]: `yarn remove axios`.

File: public/js/deleteItem.js [hunk 16]
```diff
@@ -0,0 +1,28 @@
+function deleteItem(evt, itemId, collection) {
+  evt.preventDefault();
+  console.log(itemId);
+  let message;
+
+  if (collection==='author') {
```
Comment (othman853) [inline public/js/deleteItem.js]: Faltaram uns espaços aqui. 😃

File: public/js/deleteItem.js [hunk 17]
```diff
@@ -0,0 +1,28 @@
+function deleteItem(evt, itemId, collection) {
+  evt.preventDefault();
+  console.log(itemId);
+  let message;
```
Comment (othman853) [inline public/js/deleteItem.js]: Talvez seja mais prudente usar `var` aqui. Pode ser que alguns browsers não suportem `let`.

File: public/js/deleteItem.js [hunk 18]
```diff
@@ -0,0 +1,28 @@
+function deleteItem(evt, itemId, collection) {
+  evt.preventDefault();
```
Comment (othman853) [inline public/js/deleteItem.js]: Existe algum motivo específico para evitar o evento padrão do botão?

File: src/routes/deleteItem.js [hunk 19]
```diff
@@ -0,0 +1,23 @@
+const Controller = require('../mappers/models_controller')
+
+const deleteItem = Author => (req, res, next) => {
+  const authorController = new Controller(Author)
+  if(req.body.collection === 'author'){
```
Comment (othman853) [inline src/routes/deleteItem.js]: Acredito que deveria existir uma rota de `delete` para cada tipo de recurso. Essa rota tende a ficar imensa.

File: src/routes/deleteItem.js [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+    authorController
+    .deleteAuthor(req.body.id)
```
Comment (othman853) [inline src/routes/deleteItem.js]: Não se está tratando a situação onde o `delete` dá certo. Porque? O que acontece nesse caso? Talvez a requisição Ajax nunca seja respondida? Talvez fosse interessante responder somente com um `200` para dizer que deu tudo certo:

```javascript
delete()
  .then(() => res.sendStatus(200))
```

File: src/routes/newAuthor.js [hunk 21]
```diff
@@ -0,0 +1,9 @@
+const Controller = require('../mappers/models_controller')
+
+const newAuthor = Author => (req, res, next) => {
+  const authorController = new Controller(Author)
```
Comment (othman853) [inline src/routes/newAuthor.js]: Porque isso é necessário?

File: public/js/deleteItem.js [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+  if (collection==='author') {
+    message = 'Esta operação deletará todos os cursos deste autor. Você tem certeza disto?'
+  }else if (collection==='courses') {
+    message = 'Esta operação deletará todos as aulas deste curso. Você tem certeza disto?'
+  }
+
+  if (confirm(message)) {
+    axios({
+      method:'post',
+      url:'/deleteItem/'+itemId,
```
Comment (othman853) [inline public/js/deleteItem.js]: Porque o `id` está sendo enviado tanto na URL quanto no body da requisição?

File: src/routes/deleteItem.js (hunk 19 above)
Comment (JessicaFranke) [reply src/routes/deleteItem.js]: Acredito que dentro do pedido( um painel crud simples), a rota não irá se estender, já que podera deletar apenas três collections( Author, Course, Content), caso necessário mudaremos no futuro.

Comment (othman853) [reply src/routes/deleteItem.js]: Beleza 👍
